### ISSUE TITLE
Support led synchorization

### ISSUE THREAD

**Issue by FFY00** [libratbag/libratbag#887]
Several vendors support synchronizing the leds across several devices. As libratbag, we are in a unique position to provide led synchronization across several vendors. I would really like to have support for this.

I have given some thought to this and the implementation does not appear to be that hard.

Basically we export a `SyncPools` array, a pool would be a list of devices we want to sync. Then we ask each driver to sync the device to a specific tick or to reset the effect loop, depending on the capabilities.

Logitech devices can be set to a specific tick in the effect loop and IIRC they can also report where in the loop they are. Steelseries does not support anything of sorts but we can reset loop, same goes for almost all other vendors.

Examples:
(describing what happens when we add a device to the pool, the last device is the one added)
 - Syncing a Logitech device + Logitech device: We ask the device in the pool for the tick of the effect loop and set the new device to that tick
 - Syncing a Logitech device + Steelseries device: We ask the Logitech device for the tick of the effect loop and wait for the loop to reset, when the loop resets we reset the loop of the Steelseries device
 - Syncing a Steelseries device + Steelseries device: Every time we add a device to the pool we reset the loop for all devices

I understand this might be a bit confusing but that's probably just my crappy explanation.

If only I work on this I would give an ETA of 1~2 years for the feature to arrive, though it could be sooner depending on my free time. I am not making this one of my top priorities.

cc @bentiss @whot

**Comment by FFY00** [comment]
For now I just want to get some discussion going to see if this is something we would want to do.

**Comment by johnwilliamsdigital** [comment]
I'm just trying to get my M705, MX Master, or G602 to work in the app on Kubuntu 19.10

I've got the Piper GUI but not sure I've got libratbag installed/connected to Piper correctly. 

I'm certainly a noob with this compared to the devs, but would live an easier way to plug and play to assign the thumb buttons.

**Comment by FFY00** [comment]
>I'm just trying to get my M705, MX Master, or G602 to work in the app on Kubuntu 19.10
>
>I've got the Piper GUI but not sure I've got libratbag installed/connected to Piper correctly.
>
>I'm certainly a noob with this compared to the devs, but would live an easier way to plug and play to assign the thumb buttons.

The buttons for all of those 3 mice are configured in the same way, which seems to be broken right now. Please try to comment on the related issues, what is beeing discussed here is completely unrelated.

**Comment by whot** [comment]
how many users do you think we'll have for that feature and is it worth the effort?

also, wouldn't this be better implemented as some outside tool that just pokes the dbus API at the right times? That way you can offload all the incompatibilities between devices, the configuration, etc. to that tool that can move independently of libratbag.

and, as a non-gamer - what *is* the use-case for this other than demo videos?

**Comment by FFY00** [comment]
> how many users do you think we'll have for that feature and is it worth the effort?

I think a lot of users will find it useful. I also think this will motivate people to contribute to the project, even if only with drivers that use other projects as backend (ex. openrazer).

> also, wouldn't this be better implemented as some outside tool that just pokes the dbus API at the right times? That way you can offload all the incompatibilities between devices, the configuration, etc. to that tool that can move independently of libratbag.

Well... I don't think so. We would need to export the API for getting/setting the tick in the effect cycle which I am reluctant of. I understand you want to push this away from libratbag but I think architecturally this is where it makes the most sense to implement. It is totally valid for use to rule this out as something we don't want to maintain but I would personally like to try having this feature.

> and, as a non-gamer - what is the use-case for this other than demo videos?

It does not have anything to do with gamers really. This is just an aesthetic feature. Right now I have a G513 (keyboard), a Powerplay (wireless charging gamepad + receiver) and a G Pro Wireless (mouse) and all of them have RGB lights, it would be really neat to have them all synced up. It comes down to personal preference, if you like RGB or not. One could even go a step further and get some G560 speakers and have the whole desk light up.

The main thing here would be that we are vendor agnostic, so it would work across all vendors, which I think it's amazing.

**Comment by whot** [comment]
> I also think this will motivate people to contribute to the project, 

I accept being a pessimist, but it's hard enough to get contributors to the project to simply make their mouse *work*. I don't think fancy LED effects are going to change this. Without you (and more recently @stephanlachnit) we'd still be a the "count committers on one hands" stage.

> I think architecturally this is where it makes the most sense to implement

fwiw, sometimes "good enough"  is better than perfect. I'd argue that implementing it outside first and waiting for users to be unhappy with the tick precision is less effort than going for libratbag immediately. Doubly so because you [1] may not know what you really want until you figure out that what you implemented wasn't it :) If you do find out that it doesn't scale, or work, or it needs a completely different approach you can iterate much faster before settling down on an implementation (which may potentially be inside libratbag, who knows).

[1] the generic "you", not you specifically.

**Comment by gimbas** [comment]
I like the idea, and I think it makes some sense to implement it in libratbag, though I'm not that familiar with inner workings of it.
Not sure how much value my opinion holds.

**Comment by FFY00** [comment]
> I accept being a pessimist, but it's hard enough to get contributors to the project to simply make their mouse work. I don't think fancy LED effects are going to change this. Without you (and more recently @stephanlachnit) we'd still be a the "count committers on one hands" stage.

That is true. I think, however, this will motivate people from other projects to write libratbag drivers that use other projects as a backend. But your position is totally valid, we won't be sure until someone actually comes and does that, which could never happen.

> fwiw, sometimes "good enough" is better than perfect. I'd argue that implementing it outside first and waiting for users to be unhappy with the tick precision is less effort than going for libratbag immediately. Doubly so because you [1] may not know what you really want until you figure out that what you implemented wasn't it :) If you do find out that it doesn't scale, or work, or it needs a completely different approach you can iterate much faster before settling down on an implementation (which may potentially be inside libratbag, who knows).

The thing is that we need to export a certain API to be able to implement this outside, so we have the same problem. I see two solutions. One would be that I would fork {libratbag,piper} to {libratbag,piper}-ledsync but I don't think we would get a lot of feedback. The second one would be maintaining a `ledsync` branch which would hold the ledsync patches and be rebased on master on new changes, making the ledsync patches apply cleanly on master and so, easy to chop down. We would then release from this branch, we could mark the feature as experimental. None of the solutions is great but we really don't have much choice for getting user feedback.

**Comment by gimbas** [comment]
Theres a point to be made in implementing this outside of libratbag, that being, there are a lot of devices that support rbg lighting, and libratbag will most likely never support them, things like motherboard rbg connectors and god knows what, theres rgb in everything nowadays.
If this was a seperate it could be made to support many types of devices not only mice and keyboards.

**Comment by FFY00** [comment]
Actually with the changes to support keyboards it should now be possible to support those. Previously we were forcing a device to have a resolution but now it should be possible to export a device with only leds.

**Comment by stephanlachnit** [comment]
> Theres a point to be made in implementing this outside of libratbag, that being, there are a lot of devices that support rbg lighting, and libratbag will most likely never support them, things like motherboard rbg connectors and god knows what, theres rgb in everything nowadays.
> If this was a seperate it could be made to support many types of devices not only mice and keyboards.

Had exactly the same thought, something like `librgbled`. (Maybe with a DBus interface?) When you think about rgb (& synchronization), I think it's important to think open (all kinds of hardware), and `libratbag` is the wrong place imho, since it's clearly more focused on input than on rgb.

I would help building the new library, but it's probably going to be way harder than just implementing this in `libratbag`. On the other hand we have some opportunities: there are already some games/applications on Windows which can change the rgb behavior, but it's very vendor-specific. Having a separate library can provide these features without cluttering `libratabg`, which can just take the features it needs. Would also allow custom rgb implementations with Arduinos for example.

**Comment by stephanlachnit** [comment]
Other idea, what about we try to add an interface for [OpenRGB](https://gitlab.com/CalcProgrammer1/OpenRGB)? Haven't looked into it but maybe we should orientate us on their API.

**Comment by FFY00** [comment]
I looked at it and it does not seem great.

I think we should finish #629 and then think about maybe adding a new experimental API for led synchronization.

**Comment by stephanlachnit** [comment]
Yeah discussed it on their [gitlab](https://gitlab.com/CalcProgrammer1/OpenRGB/issues/74), and they don't plan to support synchronization in the way we would. But I think since OpenRGB exists, we can drop the ideas of an own library since there is already one which has way more supported devices.

**Comment by FFY00** [comment]
After having a stable API we can see if they are interested in contributing with drivers for devices they support.
